src/main/java/com/google/sps/servlets/GetNearbyVendorsServlet.java

```diff
@@ -0,0 +1,145 @@
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    } catch (IllegalArgumentException e) {
+      System.out.println("Latitude and/or longitude outside legal range: " + e);
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+    // ---------------------------------------------------------------------------------
+    
+    // Check values exist and are in the range
+    if (prefixGeoHash.isEmpty() || distance > MAX_DISTANCE || distance < MIN_DISTANCE) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    // Generate query
+    Query query = buildGeoQuery(prefixGeoHash, hasDelivery);
+    
+    // Fetch results
+    Iterable<Entity> vendorsRetrieved = fetchVendors(query);
+
+    // Create list and filter by distance
+    List<Vendor> nearbyVendors = createVendorsList(vendorsRetrieved, clientLocation, distance);
+
+    Gson gson = new Gson();
+
+    response.setContentType("application/json;");
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(gson.toJson(nearbyVendors));
+  }
+
+  /** 
+   * Returns a query for a Vendor with a filter to match a substring (the prefix of a geoHash) and
+   * if the vendor has or not delivery service 
+   */
+  private Query buildGeoQuery(String prefixGeoHash, boolean hasDelivery) {
+    Filter lowerBound = new FilterPredicate("saleCard.location.geoHash",
+        FilterOperator.GREATER_THAN_OR_EQUAL, prefixGeoHash);
+    Filter upperBound = new FilterPredicate("saleCard.location.geoHash",
+        FilterOperator.LESS_THAN, prefixGeoHash + "\ufffd");
+    Filter deliveryFilter = new FilterPredicate("saleCard.hasDelivery",
+        FilterOperator.EQUAL, hasDelivery);
+
+    Filter geoFilter = CompositeFilterOperator.and(lowerBound, upperBound, deliveryFilter);
+
+    return new Query("Vendor").setFilter(geoFilter);
+  }
+
+  /** Returns an iterable with the results retrieved from Datastore */
+  private Iterable<Entity> fetchVendors(Query query) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    return datastore.prepare(query).asIterable();
+  }
+
+  /** Returns a list with the retrieved vendors and check they are within the requested distance*/
+  private List<Vendor> createVendorsList(Iterable<Entity> vendors, GeoPt clientLocation, float distanceLimit) {
+    List<Vendor> nearbyVendors = new ArrayList<>();
+    for (Entity vendorEntity : vendors) {
+      Vendor vendor = new Vendor(vendorEntity);
+      GeoPt vendorLocation = vendor.getBusinessInfo().getLocation().getSalePoint();
+      if (computeHaversine(clientLocation, vendorLocation) <= distanceLimit) {
+        nearbyVendors.add(vendor);
+      }    
+    }
+
+    return nearbyVendors;
+  }
+
+  /** Computes the distance between two geographical points using Haversine formula */
+  private float computeHaversine(GeoPt pointA, GeoPt pointB) {
```

Comment:
What do A and B stand for? Are they the locations of the vendor and customer?

Reply:
They do not stand for anything. The method is used to calculate the distance between any two points, hence A and B. 
We use it to calculate the distance between vendor and customer, yes.